### PR TITLE
SE: Caching: Cache "GetHashCode" for SymbolicValue

### PR DESCRIPTION
**Pull request by martin-strecker-sonarsource**
Part of #6964

### CODE REVIEW

File: analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/SymbolicValue.cs [hunk 1]
```diff
@@ -37,6 +39,11 @@ public sealed record SymbolicValue
         public IEnumerable<SymbolicConstraint> AllConstraints =>
             Constraints.Values;
 
+        public SymbolicValue()
+        {
+            hashCode = new(() => HashCode.DictionaryContentHash(Constraints), LazyThreadSafetyMode.ExecutionAndPublication);
```
Comment (martin-strecker-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/SymbolicValue.cs:44]: For `LazyThreadSafetyMode` see https://learn.microsoft.com/en-us/dotnet/api/system.threading.lazythreadsafetymode?view=net-8.0

**Comment by martin-strecker-sonarsource** [comment]
> Did you create a benchmark too?

No. This is done with the [AnalyzerRunner](https://github.com/dotnet/roslyn/tree/main/src/Tools/AnalyzerRunner) and the Visual Studio diagnostics tool. I may add a markdown file to our docs on how to do that. It might come in handy as we have quite some of these optimizations to do and this helps with guiding in the right direction.

File: analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/SymbolicValue.cs [hunk 2]
```diff
@@ -35,8 +35,22 @@ public sealed record SymbolicValue
...
+            get => constraints;
+            init
+            {
+                if (constraints != value)
+                {
+                    constraints = value;
+                    constraintsHashCode = null;
+                }
+            }
+        }
```
Comment (martin-strecker-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/SymbolicValue.cs:53]: There is also 
https://github.com/SonarSource/sonar-dotnet/blob/f3c2ea1f0dcbdf8731ecd081325a558932c91561/analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/ProgramState.cs#L138-L143

and I want to use the same pattern there as well. There it is more important to reset the cached hashcode in the init setter. There we can do this, to make the declaration more compact:

```cs
private readonly ImmutableDictionary<IOperation, SymbolicValue> operationValue;
private int? operationValueHashCode;

private ImmutableDictionary<IOperation, SymbolicValue> OperationValue
{
    get => operationValue;
    init => SetHashCodeCachedField(value, ref operationValue, ref operationValueHashCode);
}

private static void SetHashCodeCachedField<T>(T value, ref T backingField, ref int? backingFieldHashCode) where T: class
{
    if (value == backingField)
    {
        backingField = value;
        backingFieldHashCode = null;
    }
}
```

File: analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/SymbolicValue.cs [hunk 3]
```diff
@@ -35,8 +35,22 @@ public sealed record SymbolicValue
         public static readonly SymbolicValue True = NotNull.WithConstraint(BoolConstraint.True);
         public static readonly SymbolicValue False = NotNull.WithConstraint(BoolConstraint.False);
 
+        private readonly ImmutableDictionary<Type, SymbolicConstraint> constraints = ImmutableDictionary<Type, SymbolicConstraint>.Empty;
+        private int? constraintsHashCode;
```
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/SymbolicValue.cs:39]: It is and will be hash code for the entire class. No matter how much mess we add.
```suggestion
        private int? hashCode;
```

File: analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/SymbolicValue.cs (hunk 2 above)
Comment (pavel-mikula-sonarsource) [reply analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/SymbolicValue.cs:53]: The `init` and `ref` trick sounds good in principle. 

`backingFieldHashCode` should not be a parameter, the method can just reset it directly.

File: analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/SymbolicValueTest.cs [hunk 4]
```diff
@@ -335,5 +335,47 @@ public void RemoveEntry_Miss_Returns_Instance_Type()
...
         }
+
+        [TestMethod]
+        public void GetHashCode_ReturnsDifferentValuesForPredefinedValues()
+        {
+            SymbolicValue.Empty.GetHashCode().Should().Be(SymbolicValue.Empty.GetHashCode());
+            SymbolicValue.Empty.GetHashCode().Should()
+                .NotBe(SymbolicValue.Null.GetHashCode()).And
+                .NotBe(SymbolicValue.NotNull.GetHashCode()).And
+                .NotBe(SymbolicValue.True.GetHashCode()).And
```
Comment (pavel-mikula-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/SymbolicValueTest.cs:346]: Let's put the `.And` at the beginning of the line. LIke `&&`

**Review by pavel-mikula-sonarsource — APPROVED** [review]
LGTM with a small rename before merging

File: analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/SymbolicValue.cs [hunk 5]
```diff
@@ -35,8 +35,22 @@ public sealed record SymbolicValue
...
+        private int? hashCode;
+
         // SymbolicValue can have only one constraint instance of specific type at a time
-        private ImmutableDictionary<Type, SymbolicConstraint> Constraints { get; init; } = ImmutableDictionary<Type, SymbolicConstraint>.Empty;
+        private ImmutableDictionary<Type, SymbolicConstraint> Constraints
+        {
+            get => constraints;
+            init
+            {
+                if (constraints != value)
```
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/SymbolicValue.cs:47]: This condition is not covered when values are the same. We don't use it that way. And we actually shouldn't, as we should attempt no-operation when possible.

Do we really need the condition then?

Comment (martin-strecker-sonarsource) [reply analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/SymbolicValue.cs:47]: It doesn't hurt to keep it, IMO. It allows doing something like, e.g. `with Constraints = someCond ? Empty : Constraints`. I could add a test case to increase test coverage if that is a concern.

Comment (pavel-mikula-sonarsource) [reply analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/SymbolicValue.cs:47]: I would remove the condition. It's one more additional `Equals` check on a hot path that doesn't need to be there. 

And so far, we didn't need it. And it's against the intended usage of the property. We can always add it back if we need to.

Comment (martin-strecker-sonarsource) [reply analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/SymbolicValue.cs:47]: That's fine with me.

File: analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/SymbolicValue.cs [hunk 6]
```diff
@@ -35,8 +35,22 @@ public sealed record SymbolicValue
...
+                {
+                    constraints = value;
+                    hashCode = null;
+                }
```
Comment (martin-strecker-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/SymbolicValue.cs:51]:
```suggestion
                constraints = value;
                hashCode = null;
```

**Pull request merged**